File: app/src/main/java/com/popgamma/tutor/FallbackReplies.kt
package com.popgamma.tutor

// Pure Kotlin -- no Android imports. Canned replies used only when the live Gemini chat call
// fails (quota exhausted, network down, timeout). Mirrors the "canned but honest" approach the
// plan already applies to voice transcription (OfflineFixtures) -- rather than a dead-ended
// transcript with an error banner and no reply, the demo degrades to fixed text and says so.
//
// The text still has to respect the two things this whole demo argues for, or a fallback reply
// during a live pitch would undercut the argument instead of just working around an outage:
//   - structure/encouragement follow the active ToneProfile (Caveat 2)
//   - encouragement never changes with confidence routing, only follow-up depth does (Caveat 3)
object FallbackReplies {

    private const val ENCOURAGEMENT_STANDARD = "Good effort."
    private const val ENCOURAGEMENT_HIGH = "Good, and great work -- you're doing well, keep it up."

    private const val OFFLINE_NOTE = "\n\n(Offline reply -- live tutor unavailable right now.)"

    fun reply(profile: ToneProfile, correct: Boolean? = null, band: ConfidenceBand? = null): String {
        val encouragement = when (profile.encouragement) {
            EncouragementLevel.STANDARD -> ENCOURAGEMENT_STANDARD
            EncouragementLevel.HIGH -> ENCOURAGEMENT_HIGH
        }
        val body = when (profile.structure) {
            StructureLevel.LOOSE ->
                "By the way, that tracks. Let's keep going at your own pace -- no need to rush this one."
            StructureLevel.MEDIUM ->
                "First, that's on the right track. Next, let's build on it a little before moving on."
            StructureLevel.TIGHT ->
                "First, note the key point. Next, check it against the question. Then, move to the next step."
        }
        val followUp = if (correct == null || band == null) "" else when {
            correct && band == ConfidenceBand.HIGH ->
                " Let's move to the next question."
            correct && band == ConfidenceBand.LOW ->
                " Walk me through how you got that, just to check."
            !correct && band == ConfidenceBand.LOW ->
                " That's a small slip, not a big deal -- here's the fix."
            else -> // !correct && band == ConfidenceBand.HIGH -- confident wrong, the strongest signal
                " Let's slow down and walk through this one directly, step by step, since it's worth understanding fully."
        }
        return "$encouragement $body$followUp$OFFLINE_NOTE"
    }
}

File: web/src/logic/fallbackReplies.ts
import type { ToneProfile, ConfidenceBand } from '../types';

export const FallbackReplies = {
  ENCOURAGEMENT_STANDARD: 'Good effort.',
  ENCOURAGEMENT_HIGH: "Good, and great work -- you're doing well, keep it up.",
  OFFLINE_NOTE: '\n\n(Offline reply -- live tutor unavailable right now.)',

  reply(profile: ToneProfile, correct?: boolean | null, band?: ConfidenceBand | null): string {
    const encouragement =
      profile.encouragement === 'HIGH'
        ? this.ENCOURAGEMENT_HIGH
        : this.ENCOURAGEMENT_STANDARD;

    // "That tracks" / "on the right track" claim the answer was fine -- said unconditionally,
    // that's exactly the false "correct/good job for a wrong answer" bug this whole demo exists
    // to prevent, just baked into the canned offline text instead of the live prompt.
    // Encouragement (above) can still praise the effort when wrong; the body must not imply the
    // answer itself was right.
    let body = '';
    switch (profile.structure) {
      case 'LOOSE':
        body = correct === false
          ? "By the way, this one needs another look. Let's keep going at your own pace -- no need to rush it."
          : "By the way, that tracks. Let's keep going at your own pace -- no need to rush this one.";
        break;
      case 'MEDIUM':
        body = correct === false
          ? "First, let's take a closer look at this one. Next, let's build on it a little before moving on."
          : "First, that's on the right track. Next, let's build on it a little before moving on.";
        break;
      case 'TIGHT':
        body = "First, note the key point. Next, check it against the question. Then, move to the next step.";
        break;
    }

    let followUp = '';
    if (correct !== undefined && correct !== null && band !== undefined && band !== null) {
      if (correct && band === 'HIGH') {
        followUp = " Let's move to the next question.";
      } else if (correct && band === 'LOW') {
        followUp = ' Walk me through how you got that, just to check.';
      } else if (!correct && band === 'LOW') {
        followUp = " That's a small slip, not a big deal -- here's the fix.";
      } else {
        // !correct && band === 'HIGH' -- confident wrong, the strongest signal
        followUp = " Let's slow down and walk through this one directly, step by step, since it's worth understanding fully.";
      }
    }

    return `${encouragement} ${body}${followUp}${this.OFFLINE_NOTE}`;
  }
};
